### PR TITLE
(GH-50) Add alias which uses environment variables from Build

### LINKED ISSUE #50: Add alias which uses environment variables from Build
Using the infrastructure provided with https://github.com/cake-contrib/Cake.Tfs/issues/29 to provide an alias which uses environment variables set by an Azure Pipelines / TFS build

### PR DESCRIPTION
**Pull request by yansklyarenko**
@pascalberger I would appreciate your review.

A couple of questions:

- As far as I can see, the number of unit tests is quite small. On the other hand, the base class `TfsPullRequestSettings` is covered pretty good. Do you think more unit tests should be added? Most of those will be almost identical copy&paste clones...

- I've renamed the VSTS to Azure DevOps in a separate commit, otherwise, the doc comments of the new stuff added become inconsistent. Is it ok? If not, please let me know and I'll revert that.

Fixes #50.

### CODE REVIEW

**Comment by pascalberger** [comment]
@yansklyarenko Thanks for your contribution 👍 It might take some time until I can properly review it as I'm in meetings today and tomorrow.

Regarding your questions:

> As far as I can see, the number of unit tests is quite small. On the other hand, the base class TfsPullRequestSettings is covered pretty good. Do you think more unit tests should be added? Most of those will be almost identical copy&paste clones...

It's true that there are no tests for the TFS integration at the moment. Long term plan is to get rid of the TFS code completely here and have everything in Cake.Tfs (that's https://github.com/cake-contrib/Cake.Issues.PullRequests.Tfs/issues/22). Some stuff like retrieving files affected by a PR can be easily ported to Cake.Tfs, in other places it's more complicated as the logic is quite complicated (everywhere where the `GitClient` is used), but I also don't want to expose the TFS API from Cake.Tfs.

> I've renamed the VSTS to Azure DevOps in a separate commit, otherwise, the doc comments of the new stuff added become inconsistent. Is it ok? If not, please let me know and I'll revert that.

Yes, makes sense. If you want you can do this also in a separate PR and probably quicker merged. It also won't be the last rename as TFS vNext also will be renamed.

**Comment by pascalberger** [comment]
@yansklyarenko Thanks for your contribution 👍 It might take some time until I can properly review it as I'm in meetings today and tomorrow.

Regarding your questions:

> As far as I can see, the number of unit tests is quite small. On the other hand, the base class TfsPullRequestSettings is covered pretty good. Do you think more unit tests should be added? Most of those will be almost identical copy&paste clones...

It's true that there are no tests for the TFS integration at the moment. Long term plan is to get rid of the TFS code completely here and have everything in Cake.Tfs (that's https://github.com/cake-contrib/Cake.Issues.PullRequests.Tfs/issues/22). Some stuff like retrieving files affected by a PR can be easily ported to Cake.Tfs, in other places it's more complicated as the logic is quite complicated (everywhere where the `GitClient` is used), but I also don't want to expose the TFS API from Cake.Tfs.

> I've renamed the VSTS to Azure DevOps in a separate commit, otherwise, the doc comments of the new stuff added become inconsistent. Is it ok? If not, please let me know and I'll revert that.

Yes, makes sense. If you want you can do this also in a separate PR and probably quicker merged. It also won't be the last rename as TFS vNext also will be renamed.

**Comment by pascalberger** [comment]
@yansklyarenko Thanks for your contribution 👍 It might take some time until I can properly review it as I'm in meetings today and tomorrow.

Regarding your questions:

> As far as I can see, the number of unit tests is quite small. On the other hand, the base class TfsPullRequestSettings is covered pretty good. Do you think more unit tests should be added? Most of those will be almost identical copy&paste clones...

It's true that there are no tests for the TFS integration at the moment. Long term plan is to get rid of the TFS code completely here and have everything in Cake.Tfs (that's https://github.com/cake-contrib/Cake.Issues.PullRequests.Tfs/issues/22). Some stuff like retrieving files affected by a PR can be easily ported to Cake.Tfs, in other places it's more complicated as the logic is quite complicated (everywhere where the `GitClient` is used), but I also don't want to expose the TFS API from Cake.Tfs.

> I've renamed the VSTS to Azure DevOps in a separate commit, otherwise, the doc comments of the new stuff added become inconsistent. Is it ok? If not, please let me know and I'll revert that.

Yes, makes sense. If you want you can do this also in a separate PR and probably quicker merged. It also won't be the last rename as TFS vNext also will be renamed.

**Comment by pascalberger** [comment]
@yansklyarenko Thanks for your contribution 👍 It might take some time until I can properly review it as I'm in meetings today and tomorrow.

Regarding your questions:

> As far as I can see, the number of unit tests is quite small. On the other hand, the base class TfsPullRequestSettings is covered pretty good. Do you think more unit tests should be added? Most of those will be almost identical copy&paste clones...

It's true that there are no tests for the TFS integration at the moment. Long term plan is to get rid of the TFS code completely here and have everything in Cake.Tfs (that's https://github.com/cake-contrib/Cake.Issues.PullRequests.Tfs/issues/22). Some stuff like retrieving files affected by a PR can be easily ported to Cake.Tfs, in other places it's more complicated as the logic is quite complicated (everywhere where the `GitClient` is used), but I also don't want to expose the TFS API from Cake.Tfs.

> I've renamed the VSTS to Azure DevOps in a separate commit, otherwise, the doc comments of the new stuff added become inconsistent. Is it ok? If not, please let me know and I'll revert that.

Yes, makes sense. If you want you can do this also in a separate PR and probably quicker merged. It also won't be the last rename as TFS vNext also will be renamed.

**Comment by yansklyarenko** [comment]
@pascalberger Ok, I've reverted the rename commit here and issued a separate PR for that: #53.  

> Long term plan is to get rid of the TFS code completely here and have everything in Cake.Tfs

I'll take #22 next then. I'll start with the obvious easy stuff and then we can take it from there.

**Comment by yansklyarenko** [comment]
@pascalberger Hmm, it has just occurred to me that I might be starting from the wrong end. Perhaps, migrating from TFS API assemblies to the REST API calls (which is [this one](https://github.com/cake-contrib/Cake.Tfs/issues/13)) is the main enabler?

- As soon as Cake.Tfs does not depend on TFS client libraries, it will mean it doesn't use `GitHttpClient`
- As a result those methods in Cake.Issues.PullRequests.Tfs, which are now using `GitHttpClient`, could be modified to become pure calls to some Cake.Tfs API instead
- Then, `TfsPullRequestSystem` class could be refactored to avoid TFS API usage and only leverage Cake.Tfs addin API 

I might be missing the point. What do you have in your mind? If it's too many letters to type, perhaps couple of slides might present a clearer picture? 

Thanks.

**Comment by pascalberger** [comment]
@yansklyarenko Some historical context: Cake.Tfs was split out of Cake.Issues.PullRequests.Tfs, which is the successor of Cake.Prca.PullRequests.Tfs. So some issues have some history.

The main issue with rewriting to use the REST API directly is that one needs to rewrite the whole authentication stuff. The main reason for the REST API issue was cross-plattform compatibility, but it seems like the newer preview versions of the API target .NET standard which would solve this issue, altough I haven't tried them yet. On the other side there's currently no good solution I'm aware of to embedd the dependencies in assemblies targetting .NET standard, which could introduce some issues (but as there already is a dependency between Cake.Issues.PullRequests.Tfs and Cake.Issues which needs to match, this might not be that big of an issue). So I'm no longer sure the rewrite to direct REST API calls is worth the effort. 

I would propose to proceed with your first proposition: To first move the code which easily can be abstracted from Cake.Issues.PullRequests.Tfs to Cake.Tfs, and then see what code remains here and whats the best way to handle it (keep it here, add GitHttpClient to public API in Cake.Tfs, make GitHttpClient internal and available to Cake.PullRequests.Tfs, move code to Cake.Tfs and provide callbacks which give the necessary flexibility, ...)

What do you think?

File: src/Cake.Issues.PullRequests.Tfs/TfsPullRequestSystemAliases.PullRequestSystem.cs [hunk 1]
```diff
@@ -99,6 +99,69 @@ public static partial class TfsPullRequestSystemAliases
             return context.TfsPullRequests(new TfsPullRequestSystemSettings(repositoryUrl, pullRequestId, credentials));
         }
 
+        /// <summary>
+        /// Gets an object for writing issues to Team Foundation Server or Azure DevOps pull request
+        /// where all required data is taken from the appropriate system environment variable.
```
Comment (pascalberger) [inline src/Cake.Issues.PullRequests.Tfs/TfsPullRequestSystemAliases.PullRequestSystem.cs]: I would mention here that it looks for the environment variables set by Azure DevOps / TFS build:

```
... is taken from the environment variables set by the Azure Pipelines / TFS build.
```

File: src/Cake.Issues.PullRequests.Tfs/TfsPullRequestSystemAliases.PullRequestSystem.cs [hunk 2]
```diff
@@ -99,6 +99,69 @@ public static partial class TfsPullRequestSystemAliases
...
+        {
+            context.NotNull(nameof(context));
+            credentials.NotNull(nameof(credentials));
+
+            return context.TfsPullRequests(new TfsPullRequestSystemSettings(credentials));
+        }
+
+        /// <summary>
+        /// Gets an object for writing issues to Team Foundation Server or Azure DevOps pull request
+        /// where all required data (including authentication token) is taken from the appropriate system environment variable.
```
Comment (pascalberger) [inline src/Cake.Issues.PullRequests.Tfs/TfsPullRequestSystemAliases.PullRequestSystem.cs]: I would mention here that it looks for the environment variables set by Azure DevOps / TFS build:

```
... is taken from the environment variables set by the Azure Pipelines / TFS build.
```

File: src/Cake.Issues.PullRequests.Tfs/TfsPullRequestSystemAliases.PullRequestSystem.cs (hunk 1 above)
Comment (yansklyarenko) [reply src/Cake.Issues.PullRequests.Tfs/TfsPullRequestSystemAliases.PullRequestSystem.cs]: Fixed.

File: src/Cake.Issues.PullRequests.Tfs/TfsPullRequestSystemAliases.PullRequestSystem.cs (hunk 2 above)
Comment (yansklyarenko) [reply src/Cake.Issues.PullRequests.Tfs/TfsPullRequestSystemAliases.PullRequestSystem.cs]: Fixed.

**Comment by yansklyarenko** [comment]
> I would propose to proceed with your first proposition: To first move the code which easily can be abstracted from Cake.Issues.PullRequests.Tfs to Cake.Tfs, and then see what code remains here and whats the best way to handle it

Thanks for sharing the historical background. It makes sense to me.

Just FYI: I'll be on *offline* vacation the coming week. As soon as I'm back, I'll pick it up from here the way you suggested. I find the projects quite interesting and I'd like to contribute more.

Thanks for your cooperation and thorough review so far. I really appreciate your patience and detailed explanations!

**Comment by pascalberger** [comment]
@yansklyarenko Thanks for the fixes it has been merged now 👍 

Just let me know in which parts you're interested. I've tons of ideas for Cake addins, especially for the Cake.Issues set of addins (from additional linters, additional pr integration and build server support, additional reports, but also additional integration in Cake.Issues.Recipe) 😄 

Enjoy your vacation 🍹

**Pull request merged**